polyxsim/make_image.py

```diff
@@ -1,251 +1,286 @@
-
-            # 2D Gaussian on detector                   
-            if peakshape[0] == 1:
-                frame = ndimage.gaussian_filter(frame,peakshape[1]*0.5)
-            # add background
-            if self.graindata.param['bg'] > 0:
-                frame = frame + self.graindata.param['bg']*n.ones((framedimy,framedimz))
-            # add noise
-            if self.graindata.param['noise'] != 0:
-                frame = n.random.poisson(frame)
-            # apply psf
-            if self.graindata.param['psf'] != 0:
-                frame = ndimage.gaussian_filter(frame,self.graindata.param['psf']*0.5)
-            # resize, convert to integers and flip to same orientation as experimental frames
-            frame = frame[frame_add:framedimy-frame_add,frame_add:framedimz-frame_add]
-
-                    # limit values above 16 bit to be 16bit
-            frame = n.clip(frame,0,2**16-1)
-                        # convert to integers
-            frame = n.uint16(frame)
-
-                #flip detector orientation according to input: o11, o12, o21, o22
-            frame = detector.trans_orientation(frame,
-                               self.graindata.param['o11'],
-                               self.graindata.param['o12'],
-                               self.graindata.param['o21'],
-                               self.graindata.param['o22'],
-                               'inverse')
-            # Output frames 
-            if '.edf' in self.graindata.param['output']:
-                self.write_edf(i,frame)
-            if '.edf.gz' in self.graindata.param['output']:
-                self.write_edf(i,frame,usegzip=True)
-            if '.tif' in self.graindata.param['output']:
-                self.write_tif(i,frame)
-            if '.tif16bit' in self.graindata.param['output']:
-                self.write_tif16bit(i,frame)
-            print('\rDone frame %i took %8f s' %(i+1,time.clock()-t1), end=' ')
-            sys.stdout.flush()
-                
-    def write_edf(self,framenumber,frame,usegzip=False):
-        e=edfimage.edfimage()
-        e.data=frame
-        edim2,edim1=frame.shape
+    def make_images(self):
+        """
+        makeimage script produces edf diffraction images using the reflection information
+        
+                Henning Osholm Sorensen, June 23, 2006.
+        python translation Jette Oddershede, March 31, 2008
+        """
+        make_image_start = time.time()
+        for i in self.no_frames:
+            self.make_one_frame(i, is_multi=False)
+        make_image_end = time.time()
+        time_taken = make_image_end - make_image_start
+        print("\nFinished making images. Took", time_taken, "seconds")
+
+    def make_images_multi(self):
+        """
+        Multithreaded version of self.make_images
+        """
+        from multiprocessing import Pool
```

Comment:
multiprocessing was imported above?

Reply:
Other than that it looks promising to me. A few things to check:
- Can it run under python27 still ?
- Does it work for both fork and spawn based processes (e.g. linux vs mac+windows) ?
- Could there be a simple testcase script that runs some cases in test/ ands checks the results do not change ? For example, the md5 hashes for the created frames match each other.

If you are confident we can merge now and fix a problem if someone finds one

Thanks!!!